test_fms/random_numbers/test_random_numbers.F90

```diff
@@ -0,0 +1,279 @@
+  integer :: now_values(8) !> Values returned by the date_and_time() intrinsic
+
+  call set_calendar_type(JULIAN)
+  call date_and_time(values=now_values)
+
+  now = set_date(now_values(1), now_values(2), now_values(3), &
+                 now_values(5), now_values(6), now_values(7))
+end subroutine set_time
+
+!> Test random_numbers_mod's constructSeed()
+subroutine test_constructSeed
+  if (mpp_pe() .eq. 0) then
+    call constructSeed_assert(seeds(0), seeds(0), now)
+    call constructSeed_assert(seeds(0), seeds(0), now, 0)
+
+    call constructSeed_assert(seeds(1), seeds(2), now)
+    call constructSeed_assert(seeds(1), seeds(2), now, 0)
+
+    call constructSeed_assert(seeds(0), seeds(0), now,  10)
+    call constructSeed_assert(seeds(0), seeds(0), now, -10)
+
+    call constructSeed_assert(seeds(1), seeds(2), now,  10)
+    call constructSeed_assert(seeds(1), seeds(2), now, -10)
+
+    call constructSeed_assert(seeds(0), seeds(0), now,  20)
+    call constructSeed_assert(seeds(0), seeds(0), now, -20)
+
+    call constructSeed_assert(seeds(1), seeds(2), now,  20)
+    call constructSeed_assert(seeds(1), seeds(2), now, -20)
+  endif
+end subroutine test_constructSeed
+
+subroutine constructSeed_assert(i, j, time, perm)
+  integer,           intent(in) :: i, j !> Seed values
+  type(time_type),   intent(in) :: time !> Time to be used in the construction of a seed vector
+  integer, optional, intent(in) :: perm !> Permutation to be applied to the seed vector
+  integer, dimension(8) :: seed_expected, seed_ret !> Expected and returned seed vectors
+  integer :: year, month, day, hour, minute, second !> Date and time values
+
+  call get_date(time, year, month, day, hour, minute, second)
+  seed_expected = [i, j, year, month, day, hour, minute, second]
+  if(present(perm)) seed_expected = ishftc(seed_expected, perm)
+
+  seed_ret = constructSeed(i, j, time, perm)
+  call array_compare_1d(seed_ret, seed_expected, "constructSeed unit test failed")
+end subroutine constructSeed_assert
+
+!> Test random_numbers_mod's getRandomNumbers()
+subroutine test_getRandomNumbers
+  type(randomNumberStream) :: stream !> Random number stream
+
+  call calc_expected_moments
+
+  select case (mpp_pe())
+    case (0)
+      stream = initializeRandomNumberStream(seeds(0))
+      call test_getRandomNumbers_dispatch(stream)
+    case (1)
+      stream = initializeRandomNumberStream(seeds(1))
+      call test_getRandomNumbers_dispatch(stream)
+    case (2)
+      stream = initializeRandomNumberStream(seeds)
+      call test_getRandomNumbers_dispatch(stream)
+    case (3)
+      stream = initializeRandomNumberStream(constructSeed(seeds(1), seeds(2), now))
+      call test_getRandomNumbers_dispatch(stream)
+    case default
+      call mpp_error(fatal, "Unexpected PE: " // string(mpp_pe()))
+  end select
+end subroutine test_getRandomNumbers
+
+! Expression for the expectation value of the i-th raw moment
+#define CALC_MOMENT_(i) (1._k / real(i + 1, k))
```

Comment:
where did this equation come from?

Reply:
The _m_-th raw moment of a distribution, by definition, is the integral of _x_^_m_ times _f(x)_, where _f(x)_ is the distribution's probability density function. For the uniform distribution, _f(x)_=1 for 0<_x_<1, and _f(x)_=0 outside of that range. So the _m_-th moment is the integral from 0 to 1 of _x_^_m_, which is 1/(_m_ + 1).

Reply:
It's been a while since I thought about this stuff.  Why is it a uniform distribution?

Reply:
The comments in `random_number_mod` don't seem to explicitly state that the distribution is uniform. But, I think the term "random number generator" usually tends to mean "uniform random number generator", unless some other distribution is explicitly stated. The [0, 1] output range is also a hint that it's intended to output uniform random numbers.